Mobile: Add list continuation & Markdown preview to editor

devonzuegel opened the pull request:
Discourse discussions:
- Resolves https://discourse.joplinapp.org/t/feature-request-list-continuation-on-mobile/1742/17
- Moves towards a WYSIWYG editor a la https://discourse.joplinapp.org/t/wysiwyg-editor-in-joplin/2253/73 (though it's still clearly not a real WYSWYG editor)

Demo: 

![Kapture 2019-12-25 at 22 59 53](https://user-images.githubusercontent.com/6979755/71453521-656ed380-276a-11ea-8998-4bdfacd9896d.gif)


Next steps before merging:
- [x] Fix the keyboard issue (`⌘K`)
- [x] Fix the checkbox issue
- [ ] Handle TODOs
- [x] I still haven't been able to successfully set up Android on my machine, so I'd love for someone to test this on theirs and report back before we merge!

Next steps after this PR:
- [ ] Support custom theme in the Markdown preview
- [ ] Consider removing the "Note view" screen that comes before the editor so that it goes straight to the editor. When I open a note, I'm more likely to want to go to the edit screen so I can edit it rapidly, rather than to see how it looks rendered. Plus, the edit screen 
- [ ] Support tabs to indent and unindent

devonzuegel commented:
Hi @laurent22! This is still a draft PR, ~80% of the way done. Before I polish it up and ask for a deep dive review, would you prefer I separate out the PRs for the list continuation and Markdown preview?

devonzuegel commented:
Friendly bump! (No pressure at all if you've already seen it, I'm just not sure if Draft PRs send notifications.)

laurent22 commented on ReactNativeClient/lib/components/screens/note.js:
```diff
@@ -316,7 +322,7 @@ class NoteScreenComponent extends BaseScreenComponent {
 	}
 
-	body_changeText(text) {
+	saveText(text) {
```
Why rename this?

laurent22 commented on ReactNativeClient/lib/components/screens/note.js:
```diff
@@ -750,7 +756,6 @@ class NoteScreenComponent extends BaseScreenComponent {
 		if (this.state.mode === 'view') fieldToFocus = '';
 
 		if (fieldToFocus === 'title') this.refs.titleTextField.focus();
-		if (fieldToFocus === 'body') this.refs.noteBodyTextField.focus();
```
Any reason this was removed? This is to focus the right field (either title or body) depending on user settings.

laurent22 commented:
Thanks @devonzuegel, it looks very good as far as I can see, although I didn't review in details the MarkdownEditor files. I've left two comments on the code, just minor things.

One more important one is that I would prefer if the MarkdownEditor folder was only part of the ReactNativeClient directory - i.e. if it was outside of `lib`. These files are for the mobile app only so don't need to be shared with all the other clients. I know that in "lib" there are some mobile-only files (for historical reasons, because the mobile app was the first one) but for new app-specific files I prefer if they are outside of lib.

devonzuegel commented:
Thank you for the review. Sorry for the delay, it's been busy at work! I'll get this over the finish line once I have a little time.

> One more important one is that I would prefer if the MarkdownEditor folder was only part of the ReactNativeClient directory 

Cool, that makes sense. I'll look into making that change. 👍

laurent22 commented:
That's a feature that I think many people will want, so let's keep it open for now, also in case someone else wants to finish it.

gasolin commented:
@devonzuegel the code naming/structure looks pretty similar as https://github.com/kunall17/react-native-markdown-editor/blob/master/lib/src/Formats.js

can we just reuse the package, or is there some key difference between yours and https://github.com/kunall17/react-native-markdown-editor ?

devonzuegel commented:
> One more important one is that I would prefer if the MarkdownEditor folder was only part of the ReactNativeClient directory

I moved `MarkdownEditor/` up out of `ReactNativeClient/lib/` into `ReactNativeClient/`:

![image](https://user-images.githubusercontent.com/6979755/75631603-28c7ba00-5ba9-11ea-85a3-74d3571ab6ad.png)


But I don't know how to require it from `ReactNativeClient/lib/components/screens/note.js` when it's not also in `lib/`. Two things I tried:

<img width="49%" src="https://user-images.githubusercontent.com/6979755/75631654-83611600-5ba9-11ea-95e8-2de4528b950d.png" /> <img width="50%" src="https://user-images.githubusercontent.com/6979755/75631657-996ed680-5ba9-11ea-87bc-fc8089948b19.png" />

devonzuegel commented:
I managed to move most of it out of `ReactNativeClient/lib/` by having just a wrapper file in `lib/`, but I don't think that quite resolves your request @laurent22 since it still leaves a `ReactNativeClient/lib/MarkdownEditor.js` file in `lib/`. Hopefully a step in the right direction though!

laurent22 commented:
> But I don't know how to require it from ReactNativeClient/lib/components/screens/note.js when it's not also in lib/

I think you'd need to add as many ".." as needed to get to the file you want to require.

In this case, it should be ../../../MarkdownEditor

laurent22 commented:
By the way, what was the reason for not directly using react-native-markdown-editor?

devonzuegel commented:
> In this case, it should be ../../../MarkdownEditor

I tried this too, but it originally gave me an error. It's working now though! I must've had something else wrong with the build the first time I tried it.

> By the way, what was the reason for not directly using react-native-markdown-editor?

A few reasons:
- It made [a bunch of assumptions](https://github.com/kunall17/react-native-markdown-editor/blob/c29b3811da18b0bde3078975712d2464887dd60d/lib/src/MarkdownEditor.js#L17-L46) about the layout and style of the app, which didn't work appropriately for Joplin
- I wanted to have a lighter-weight set of offered formats in the menu, and to give them what I think are better names. (Compare their [14 formats](https://github.com/kunall17/react-native-markdown-editor/blob/c29b3811da18b0bde3078975712d2464887dd60d/lib/src/Formats.js#L7-L32) to the [10 formats](https://github.com/devonzuegel/joplin/blob/73df3472f6bb0f0763a09d9644990eebdf72c51a/ReactNativeClient/MarkdownEditor/Formats.js#L7-L22) I selected for Joplin's menu.)
- It appears to not be maintained – https://github.com/kunall17/react-native-markdown-editor/commits/master
- It's fairly small and simple, so it's easy to inline
- I think there may have been some behavior/interaction design choice I disagreed with in the library somewhere, but I can't remember what it is now... I'll try to track it down. :)

The key thing is it just feels like that lib was designed for a very specific purpose, not so much with general use and flexibility in mind. In other words, it's not really a "library", just some open sourced custom code. I could try to upstream my changes so that the module is more composable, but I expect it'd take a long time to get those fixes in.

laurent22 commented:
Thanks for the explanation, and indeed your approach in that case makes sense. If you hadn't copied the code directly, we probably should have forked it anyway, so we may as well put in the codebase directly as it's easier to work on it.

bedwardly-down commented:
@devonzuegel , on a side note, you have an amazing blog on top of the insane amount of work you've been putting into here. I can't wait till this gets finished. :heart:

devonzuegel commented:
In the "Next steps before merging" in the OP, I said I'd enable custom styles instead of how they're hard-coded right now. But since folks have been waiting for this for so long (sorry for the delay!), I'm thinking it might make sense to remove the TODOs, merge this as-is, and then enable custom styles in another, smaller PR.

Does that sound good? And is there anything else I should do before this is ready to merge? I believe I addressed all the comments.

laurent22 commented:
By the way, you might want to run `npm run linter ./` on the codebase to get rid of the CI errors (I've added a few new eslint rules recently).

devonzuegel commented on ReactNativeClient/MarkdownEditor/MarkdownEditor.js:
```diff
@@ -0,0 +1,171 @@
+	};
+
+	componentDidMount() {
+		this.textInput.focus();
```
Good point. I've pushed some changes that incorporate that feedback.

I'd love your input on my implementation. The way I'm referencing the grandchild refs from `note.js` feels really inelegant, and it means that the grandparent needs to know about the implementation details of not just the child but the grandchild, but I didn't have any particularly better ideas:

https://github.com/laurent22/joplin/pull/2224/files#diff-b185419b435222220c2d114eb5aba85dR780-R784

laurent22 replied:
Hmm, yes the code makes sense but it's indeed a bit convoluted. Maybe you could simply add a `focus()` method on MarkdownEditor, which would call focus() on its TextInput child component. Then you can directly call markdownEditorRef.current.focus() from the parent. Unless I missed something I think that would work. And that way, the parent doesn't need to know about the internal details of MarkdownEditor.

devonzuegel replied:
Much better approach!

devonzuegel commented on ReactNativeClient/MarkdownEditor/MarkdownEditor.js:
```diff
@@ -0,0 +1,171 @@
+	getState = () => {
+		this.setState({selection: {start: 1, end: 1}});
+		return this.state;
+	};
```
I don't remember why this is in here, and after some testing without it, it seems to have the proper behavior anyways... so I'll remove it. :)

devonzuegel commented:
> What I'm wondering is how stable this editor is for day to day use? Is it ok to completely switch to it, or should we have the previous editor and this one (marked as "Beta") side-by-side with a way to switch from one to another? This is in case we find some big issue we missed during testing. Then after a few weeks, if everything's good, we switch completely to it. What do you think?

I like this idea a lot! I'm not confident this new editor is totally stable, since I've only tested it on the iOS emulator and not on my actual phone. Plus I have no idea how it'll behave on an Android device, since I haven't been able to get the Android emulator running on my laptop.

We could have some sort of "join this beta" toggle, and perhaps even link to a Discourse thread asking for feedback/bug reports. Wdyt about that general idea, and the specific implementation?

gasolin commented:
@devonzuegel test and works fine on Android emulator 👍 

a suggestion: Change `Code` to `</>` which aligns with the Desktop's shortcut.

FYR desktop shortcut order is [ **B**, *I*, | Link, </>, attachment, | number, list, checkbox, Header, Horizontal, Datetime]

I suggest align from a shorter list with the same order [**B**, *I*, | Link, </>, | List, H(rename from `H2`) ]

And will be nice to 
1. replace text with existing desktop icons
2. add numbered list (once switched to the desktop icons)

devonzuegel commented:
All righty! Changes since your last review:
- I put it under its own beta flag, in case we want to add additional betas in the future. 
- I put it under the "Note" category, though I considered creating a separate "Beta" category. Happy to do that if preferable. 
- I also slightly tweaked the description text styles to look a little nicer.

Here's how it looks right now:

<img width="60%" src="https://user-images.githubusercontent.com/6979755/77239248-9c9c2780-6b95-11ea-8535-f9e0d534b6c7.png"/>

I believe this is about ready to be merged!

laurent22 commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -552,7 +563,15 @@ class Setting extends BaseModel {
+				section: 'sync',
+				label: () => _('Fail-safe'),
+				description: () => _('Do not wipe out local data when sync target is empty (often the result of a misconfiguration or bug)'),
+			},
```
Please set the description to "Fail-safe: Do not wipe out local data when sync target is empty (often the result of a misconfiguration or bug)", otherwise it will invalidate all the translations which I don't think is necessary.

devonzuegel replied:
Good catch, thanks!

laurent22 commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -316,6 +316,17 @@ class Setting extends BaseModel {
+				public: true,
+				appTypes: ['mobile'],
+				label: () => _('Opt-in to the editor beta'),
+				description: () => _('This beta adds list continuation, Markdown preview, and Markdown shortcuts. If you find bugs, please report them in the Discourse forum.'),
```
Please remove the `_()` markers as we don't need to translate these strings, since I assume we'll take the editor out of beta soon.

laurent22 commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -316,6 +316,17 @@ class Setting extends BaseModel {
 			'folders.sortOrder.reverse': { value: false, type: Setting.TYPE_BOOL, public: true, label: () => _('Reverse sort order'), appTypes: ['cli'] },
 			trackLocation: { value: true, type: Setting.TYPE_BOOL, section: 'note', public: true, label: () => _('Save geo-location with notes') },
+
+			editorBeta: {
```
Maybe `editor.beta` for consistency with the other `editor.something` properties

devonzuegel commented:
> Also does my comment [here](https://github.com/laurent22/joplin/pull/2224#discussion_r395683596) make sense? I'm not entirely sure it can be implemented as I've suggested actually so if that's the case, we can leave that aside.

Yes, much better way to do it! Just pushed that change, thanks for the suggestion.

devonzuegel commented:
All righty, here's a demo that incorporates the PR feedback:

<img width="50%" src="https://user-images.githubusercontent.com/6979755/77284329-a99b4280-6c8b-11ea-8830-2e18ac42b059.gif"/>

There are a lot more improvements we can make down the line. A few ideas:
- We can make it possible to expand the preview window to take more/less of the screen, rather than being static at 50% of the screen space
- We can clean up the styles, since they're not very nice right now 
- We can make the styles customizable
- We can make the preview pane fade in nicely (I started playing with this but realized I should focus and just finally get this PR done 😝)

Excited to hear folks' feedback on it!

laurent22 commented:
Thanks @devonzuegel, that looks very good, and it's nice you've re-used the NoteBodyViewer component for the preview, I hadn't noticed at first. Indeed if later we can update the button style that would be great, if only setting the colour to blue or something.

Could you fix the conflicts please (which I've unfortunately caused in my last push)? and then we can merge.

devonzuegel commented:
All righty! I've resolved merge conflicts. Here's a final gif showing the features:

<img width="50%" src="https://user-images.githubusercontent.com/6979755/77389404-5808bb80-6d50-11ea-8c42-5ca45413ce55.gif" />

I also gave in to temptation and made the styles of the Markdown format buttons conform to the theme styles. Here's how they look with different themes:
|           |                                                                                                                               |
| --------- | ----------------------------------------------------------------------------------------------------------------------------- |
| Light     | <img width="80%" src="https://user-images.githubusercontent.com/6979755/77389200-b71a0080-6d4f-11ea-8beb-72c399497f50.png" /> |
| Dark      | <img width="80%" src="https://user-images.githubusercontent.com/6979755/77389233-d44ecf00-6d4f-11ea-8f1a-88edf2d34adb.png" /> |
| OLED Dark | <img width="80%" src="https://user-images.githubusercontent.com/6979755/77389252-e892cc00-6d4f-11ea-9254-536cb6951acf.png" /> |

laurent22 commented:
Looks good, thanks for adding support for the current theme! The last thing would be to remove package-lock.json from the pull request. I thought I could fix that from GitHub so I've clicked on Delete File, but as I was afraid that had the effect of deleting the file instead of removing it from the commit. So please could you take this change out?

devonzuegel commented:
Sure thing! Done, I think.

laurent22 commented:
Perfect, many thanks for this great feature @devonzuegel! I'll try to create a release soon so that users can start testing it.

themightychris commented:
w00t! thanks @devonzuegel !!!

jotoeri commented:
Yeah! I can just ammend one more Thank You @devonzuegel ! ❤️ 
(Sorry for mail-spam here to some users, but i think it's necessary. ;) )

hemanti0503 commented:
I would love this feature so much! Thanks a lot! 😊

However, I can't find it in the configuration. Did it get removed? If yes, why? Any chance on getting it back?

laurent22 commented:
Unfortunately the React Native TextInput on which this editor is based has several issues which makes the app crash. I was hoping the TextInput component would get better over time, but it seems it got a little worse in the latest RN version (0.63), so for now I had to disable the editor.

Let's see in future RN releases, perhaps they'll eventually fix the TextInput bugs and we can re-enable the editor again.

hemanti0503 commented:
Thanks for the explanation, @laurent22. I'm keeping my fingers crossed.

HiFiJ commented:
Hi everyone,

thanks for all your great great efforts on Joplin! 
With regards to this PR, I believe this to be very important. Maybe a hidden switch could be made available to enable this editor component (albeit being crash prone, as noted above)? Editing notes is really cumbersome on android, in my case oftentimes leading to me trying to remember what I wanted to add and just do it later on a PC (... or forgetting about it).
Especially entering/editing lists without list continuation is very annoying. I'd certainly trade an improved, albeit crashing editor component with working list continuation anytime against pressing 4 keys just for beginning a single list item. 
Thanks again very much for your astonishing work,

Joost

memeplex commented:
@laurent22 are there any open issues in React Native tracker for the problems you mentioned for the editor component? I would like to track them. Thanks.

memeplex commented:
OTOH could list continuation -specially for checklists which are cumbersome to type, even with keyboard shortcuts- be added without the entire rich editing component meanwhile, or is the code too tightly coupled? A time ago I used Orgzly and they did that for checklists, it's not ideal but it's something. This keeps me using Keep for shopping lists and suchlike.

TheMrAnderson commented:
I too am curious about this.  List continuation in Android would be awesome, even if it's only certain types for a while like checkboxes.  The full feature on this looks amazing, but I can absolutely settle for plain text line continuation.  At this point I normally write items on the phone if my computer isn't handy, then update on the computer.  But when I was at a project house making a shopping list of things to grab, having just the text to copy down as we were quickly listing out the items we needed would have been incredibly helpful.

laurent22 commented:
Unfortunately even for just list continuation we need a reliable text input component, because we need to be able to position the cursor correctly. However as of now, positioning doesn't work well and can even crash the application so we'll need to wait for React Native to become more reliable.

TheMrAnderson commented:
That makes sense.  Thank you

semyl commented:
Is it possible to separate out the toolbar from the list continuation? Or is the toolbar also reliant on the cursor position part?

It would be a huge win to be able to click a button and insert a checkbox, even without list continuation. Thanks. Sorry I don't know enough to look at the code myself.
